distributed/shuffle/_comms.py

```diff
@@ -29,24 +29,23 @@ class CommShardsBuffer(ShardsBuffer):
 
     State
     -----
-    memory_limit: str
-        A maximum amount of memory to use across the process, like "1 GiB"
-        This includes both data in shards and also in network communications
-    max_connections: int
-        The maximum number of connections to have out at once
-    max_message_size: str
-        The maximum size of a single message that we want to send
+    max_message_size: int
+        The maximum size in bytes of a single message that we want to send
 
     Parameters
     ----------
-    send: callable
+    send : callable
         How to send a list of shards to a worker
         Expects an address of the target worker (string)
         and a payload of shards (list of bytes) to send to that worker
+    memory_limiter : ResourceLimiter, optional
+        Limiter for memory usage (in bytes), or None if no limiting
+        should be applied.
```

Comment:
This limiter is a bit more subtle. P2P is push based but we also needed a mechanism for backpressure control since otherwise workers may just die because they receive too much input data and are not fast enough in writing things back to disk.

Therefore, this limiter actually waits/blocks *after* data has been received until we're below this threshold.

Reply:
Slight off-topic: It is also typically process wide to protect from overflow in case multiple shuffles are running. This process wide limiting has been implemented prematurely as it turns out since most seemingly concurrent shuffles (e.g. a dataframe merge) are still executed sequentially due to task prioritization

Reply:
- When a worker is _receiving_ data it increases the limiter counter https://github.com/dask/distributed/blob/f4328bbd919a08fc351a7514bf7895e14a5de7a4/distributed/shuffle/_buffer.py#L199-L200
- If after accepting the shards it is over the limit, it blocks the transfer, i.e. the connection stays open while the push/transfer job waits for completion https://github.com/dask/distributed/blob/f4328bbd919a08fc351a7514bf7895e14a5de7a4/distributed/shuffle/_buffer.py#L206-L207
- and once a shard is processed (i.e. either submitted or written to disk) we are decreasing the counter https://github.com/dask/distributed/blob/f4328bbd919a08fc351a7514bf7895e14a5de7a4/distributed/shuffle/_buffer.py#L122-L123

Reply:
this all doesn't necessarily need to be part of the doc string but I believe what we're doing is uncommon enough that it warrants a bit of explanation what "limiting" is actually doing.
I consider it something like an _optimistic_ resource limiting rather than a strict pool/semaphore

Reply:
> Therefore, this limiter actually waits/blocks after data has been received until we're below this threshold.

This is also for diskbuffer, right?

Reply:
> This is also for diskbuffer, right?

Yes, this is true for both buffers and if either buffer is full on sender or receiver side it would block the current transfer task